calculator.go

```diff
@@ -1,13 +1,90 @@
 // Package calculator provides a library for simple calculations in Go.
 package calculator
 
+import (
+	"errors"
+	"regexp"
+	"strconv"
+)
+
+var evalRX *regexp.Regexp = regexp.MustCompile(`^\s*(\d+\.?\d*)\s*([^\d\.])\s*(\d+\.?\d*)\s*$`)
+
 // Add takes two numbers and returns the result of adding them together.
-func Add(a, b float64) float64 {
-return a + b
+func Add(a float64, b ...float64) float64 {
+	for _, bb := range b {
+		a += bb
+	}
+	return a
 }
 
 // Subtract takes two numbers and returns the result of subtracting the second
 // from the first.
-func Subtract(a, b float64) float64 {
-	return b - a
+func Subtract(a float64, b ...float64) float64 {
+	for _, bb := range b {
+		a -= bb
+	}
+	return a
+}
+
+// Multiply takes two numbers and returns the result of multiplying them
+func Multiply(a float64, b ...float64) float64 {
+	for _, bb := range b {
+		a *= bb
+	}
+	return a
+}
+
+// Divide takes two numbers and divides the first one by the second one
+func Divide(a float64, b ...float64) (float64, error) {
+	for _, bb := range b {
+		if bb == 0 {
+			return 0, errors.New("division by zero")
+		}
+		a /= bb
+	}
+	return a, nil
+}
+
+// Sqrt returns the square root of a positive number, an error otherwise.
+func Sqrt(a float64) (float64, error) {
+	if a < 0 {
+		return 0, errors.New("square root of a negative number")
+	}
+	x, y := a/2, 0.0
+	for i := 0; i < 5000; i++ {
+		if x == 0 {
+			break
+		}
+		y = 0.5 * (x + (a / x))
+		x = y
+	}
+
+	return y, nil
+}
+
+//Evaluate evaluates a given string expression and returns the result or an error
+func Evaluate(e string) (float64, error) {
+	m := evalRX.FindStringSubmatch(e)
+	if len(m) == 0 {
+		return 0, errors.New("invalid expression")
+	}
+	a, err := strconv.ParseFloat(m[1], 64)
+	if err != nil {
+		return 0, err
+	}
+	b, err := strconv.ParseFloat(m[3], 64)
+	if err != nil {
+		return 0, err
+	}
+	switch m[2] {
+	case "+":
+		return Add(a, b), nil
+	case "-":
+		return Subtract(a, b), nil
+	case "*":
+		return Multiply(a, b), nil
+	case "/":
+		return Divide(a, b)
+	}
+	return 0, errors.New("invalid operation")
```

Comment:
Let's make the 'invalid operation' error the `default` case in the switch statement. Then the function itself can return a nil error, which is more natural.

Reply:
A `return` in the `default` case would render anything after it unreachable. I've change this a bit to set `err` within `default`, and return it at the end of the function.

Reply:
Good point! I was thinking that the `return` would be mandatory, but as you say, if the default case returns, then that won't be an issue. So we could write something like:

```go
	switch x {
	...
	default:
		return 0, errors.New("bogus")
	}
}
```

Does that work?

Reply:
So it is generally acceptable to not have a return outside of a condition in functions if it's within a `switch`'s `default` case? i.e instead of having to do:
```go
func Evaluate(e string) (float64, error) {
	...
	switch op {
	...
	default:
		err = errors.New("invalid operation")
	}
	return 0, er
}
```
, it is okay to do:
```go
func Evaluate(e string) (float64, error) {
	...
	switch op {
	...
	default:
		return 0, errors.New("invalid operation")
	}
}
```
?

Reply:
Yes, that's fine! The spec only requires that a function have a _terminating statement_, which includes a switch statement whose default case ends with `return`: https://golang.org/ref/spec#Terminating_statements

Reply:
Done.